File: plugin/src/main/com/suhininalex/clones/clonefilter/SubclassFilter.java
package com.suhininalex.clones.clonefilter;

import com.suhininalex.clones.CloneClass;
import com.suhininalex.suffixtree.Node;

import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;

public class SubclassFilter implements CloneClassFilter {

    /**
     * link from node to cloneClass with suffix link to this node
     */
    private final Map<Node, CloneClass> reverseSuffixLink = new IdentityHashMap<>();

    public SubclassFilter(Iterable<CloneClass> cloneClassesToFilter){
        for (CloneClass cloneClass : cloneClassesToFilter){
            Node node = cloneClass.getTreeNode();
            if (node.getSuffixLink()!=null)
                reverseSuffixLink.put(node.getSuffixLink(), cloneClass);
        }
    }

    @Override
    public boolean isAllowed(CloneClass cloneClass) {
        CloneClass greaterClass = reverseSuffixLink.get(cloneClass.getTreeNode());
        if (greaterClass==null) return true;
        else if (greaterClass.getSize()==cloneClass.getSize())
            return false;
        else return true;
    }
}

File: plugin/src/main/com/suhininalex/clones/clonefilter/SubclassFilter.kt
package com.suhininalex.clones.clonefilter

import com.suhininalex.clones.CloneClass
import com.suhininalex.suffixtree.Node
import stream

import java.util.IdentityHashMap

class SubclassFilter(cloneClassesToFilter: Iterable<CloneClass>) : CloneClassFilter {

    /**
     * link from node to cloneClass with suffix link to this node
     */
    private val reverseSuffixLink = IdentityHashMap<Node, CloneClass>()
        .apply {
            cloneClassesToFilter.stream().filter { it.treeNode.suffixLink!=null }
                .forEach { put(it.treeNode.suffixLink, it) }
        }

    override fun isAllowed(cloneClass: CloneClass): Boolean {
        val greaterClass = reverseSuffixLink[cloneClass.treeNode] ?: return true
        return greaterClass.size != cloneClass.size
    }

}
